plugins/inputs/mcrouter/mcrouter.go

```diff
@@ -0,0 +1,273 @@
+	"retrans_per_kbyte_max":                      typeInt,
+	"cmd_get_count":                              typeInt,
+	"cmd_delete_out":                             typeInt,
+	"cmd_lease_get":                              typeInt,
+	"cmd_set":                                    typeInt,
+	"cmd_get_out_all":                            typeInt,
+	"cmd_get_out":                                typeInt,
+	"cmd_lease_set_count":                        typeInt,
+	"cmd_other_out_all":                          typeInt,
+	"cmd_lease_get_out":                          typeInt,
+	"cmd_set_count":                              typeInt,
+	"cmd_lease_set_out":                          typeInt,
+	"cmd_delete_count":                           typeInt,
+	"cmd_other":                                  typeInt,
+	"cmd_delete":                                 typeInt,
+	"cmd_get":                                    typeInt,
+	"cmd_lease_set":                              typeInt,
+	"cmd_set_out":                                typeInt,
+	"cmd_lease_get_count":                        typeInt,
+	"cmd_other_out":                              typeInt,
+	"cmd_lease_get_out_all":                      typeInt,
+	"cmd_set_out_all":                            typeInt,
+	"cmd_other_count":                            typeInt,
+	"cmd_delete_out_all":                         typeInt,
+	"cmd_lease_set_out_all":                      typeInt,
+}
+
+// SampleConfig returns sample configuration message
+func (m *Mcrouter) SampleConfig() string {
+	return sampleConfig
+}
+
+// Description returns description of Mcrouter plugin
+func (m *Mcrouter) Description() string {
+	return "Read metrics from one or many mcrouter servers"
+}
+
+// Gather reads stats from all configured servers accumulates stats
+func (m *Mcrouter) Gather(acc telegraf.Accumulator) error {
+	ctx := context.Background()
+
+	if m.Timeout.Duration < 1*time.Second {
+		m.Timeout.Duration = defaultTimeout
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, m.Timeout.Duration)
+	defer cancel()
+
+	if len(m.Servers) == 0 {
+		m.Servers = []string{defaultServerURL.String()}
+	}
+
+	for _, serverAddress := range m.Servers {
+		acc.AddError(m.gatherServer(ctx, serverAddress, acc))
+	}
+
+	return nil
+}
+
+func (m *Mcrouter) gatherServer(ctx context.Context, address string, acc telegraf.Accumulator) error {
+	var conn net.Conn
+	var err error
+	var protocol string
+	var host string
+	var port string
+	var dialer net.Dialer
+
+	u, parseError := url.Parse(address)
+
+	if parseError != nil {
+		return fmt.Errorf("Invalid server address")
+	}
+
+	if u.Scheme != "tcp" && u.Scheme != "unix" {
+		protocol = defaultServerURL.Scheme
+	} else {
+		protocol = u.Scheme
+	}
+
+	if protocol == "unix" {
+		if u.Path == "" {
+			return fmt.Errorf("Invalid unix socket path")
+		}
+
+		address = u.Path
+	} else {
+		if u.Host == "" {
+			u.Host = defaultServerURL.Host
+		}
+
+		host, port, err = net.SplitHostPort(u.Host)
+
+		if err != nil {
```

Comment:
I did a few quick tests and unfortunately I think we will need to handle the case when either host or port are empty strings:
```
func Check(input string) {
	host, port, err := net.SplitHostPort(input)
	fmt.Printf("%q %q %q: %v\n", input, host, port, err)
}

func TestSplitHostPort(t *testing.T) {
	Check("localhost:8086")
	Check("localhost") # we should accept this and use default port
	Check("localhost:") # use default port
	Check(":8086") # use default hostname
	Check(":") # default host:port or reject
	Check("") # reject, but we already do that above
}
```
```
"localhost:8086" "localhost" "8086": <nil>
"localhost" "" "": address localhost: missing port in address
"localhost:" "localhost" "": <nil>
":8086" "" "8086": <nil>
":" "" "": <nil>
"" "" "": missing port in address
```

Reply:
I added a test for parsing addresses and separated the address parsing into its own function.

I'm just using the `url.URL` functions to get the hostname and port parts and not using `net.SplitHostPort` any more.